[JSC] Implement support for class static initialization blocks

hyjorc1 opened the pull request:
#### c33391001b54c44edeaac63efff2b07c0e57e7be
<pre>
[JSC] Implement support for class static initialization blocks
<a href="https://bugs.webkit.org/show_bug.cgi?id=235085">https://bugs.webkit.org/show_bug.cgi?id=235085</a>
rdar://99056882

Reviewed by Yusuke Suzuki.

Class static initialization block is a new feature of a class to perform
additional static initialization during class definition evaluation.

```
class C {
    static { /* … */ }
}
```

TC39 Spec: <a href="https://tc39.es/proposal-class-static-block/">https://tc39.es/proposal-class-static-block/</a>
TC39 Proposal: <a href="https://github.com/tc39/proposal-class-static-block">https://github.com/tc39/proposal-class-static-block</a>
MDN Web Doc: <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_static_initialization_blocks">https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_static_initialization_blocks</a>

In this patch, static blocks are implemented as functions which are
evaluated along with the initialization of static class fields during
class definition evaluation. This can be further optimized by inlining
static block functions to the field initialization.

* JSTests/stress/class-static-block.js: Added.
(assert):
(A):
(assert.C):
(assert.B):
(assert.D):
(assert.A):
(assert.A.friendA.prototype.getX):
(assert.A.friendA.prototype.setX):
(assert.A.prototype.getX):
(assert.inner):
(catch.C.prototype.async inner):
(catch.C):
(catch):
(async inner.C.prototype.async inner):
(async inner.C):
(async inner):
(C.inner):
(C):
(await.C.inner):
(await.C):
(await):
(arguments.C.inner):
(arguments.C):
(arguments):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::PropertyListNode::emitBytecode):
(JSC::FunctionCallValueNode::emitBytecode):
(JSC::FuncExprNode::emitBytecode):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createFunctionExpr):
(JSC::ASTBuilder::createProperty):
(JSC::ASTBuilder::makeFunctionCallNode):
* Source/JavaScriptCore/parser/NodeConstructors.h:
(JSC::PropertyNode::PropertyNode):
(JSC::FunctionCallValueNode::FunctionCallValueNode):
(JSC::FuncExprNode::FuncExprNode):
* Source/JavaScriptCore/parser/Nodes.h:
(JSC::FuncExprNode::isStaticBlockFunction const):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::isArrowFunctionParameters):
(JSC::Parser&lt;LexerType&gt;::parseStatementListItem):
(JSC::Parser&lt;LexerType&gt;::parseVariableDeclarationList):
(JSC::Parser&lt;LexerType&gt;::parseBreakStatement):
(JSC::Parser&lt;LexerType&gt;::parseContinueStatement):
(JSC::Parser&lt;LexerType&gt;::parseReturnStatement):
(JSC::Parser&lt;LexerType&gt;::parseTryStatement):
(JSC::Parser&lt;LexerType&gt;::parseBlockStatement):
(JSC::stringArticleForFunctionMode):
(JSC::stringForFunctionMode):
(JSC::Parser&lt;LexerType&gt;::parseFunctionParameters):
(JSC::Parser&lt;LexerType&gt;::parseFunctionInfo):
(JSC::Parser&lt;LexerType&gt;::parseClass):
(JSC::Parser&lt;LexerType&gt;::parseClassFieldInitializerSourceElements):
(JSC::Parser&lt;LexerType&gt;::parseAssignmentExpression):
(JSC::Parser&lt;LexerType&gt;::parsePrimaryExpression):
(JSC::Parser&lt;LexerType&gt;::parseMemberExpression):
(JSC::Parser&lt;LexerType&gt;::parseUnaryExpression):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::setSourceParseMode):
(JSC::Scope::setIsStaticBlockScope):
(JSC::Scope::isStaticBlockScope):
(JSC::Parser::canUseIdentifierAwait):
(JSC::Parser::disallowedIdentifierAwaitReason):
(JSC::Parser::findClosetFunctionScope):
(JSC::Parser::findClosetAsyncFunctionScope):
(JSC::Parser::findScopeUntilStaticBlock):
* Source/JavaScriptCore/parser/ParserModes.h:
(JSC::isFunctionParseMode):
(JSC::isMethodParseMode):
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::makeFunctionCallNode):
(JSC::SyntaxChecker::createFunctionExpr):
(JSC::SyntaxChecker::createProperty):
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::toStringSlow):

Canonical link: <a href="https://commits.webkit.org/255173@main">https://commits.webkit.org/255173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7555f15ffeb15c6be7f33f859e9cedf4a94c7592

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91556 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101268 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161348 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/719 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97619 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/449 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/78248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27401 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82373 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70440 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82993 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35681 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16055 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78037 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33452 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17149 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26951 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3589 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37274 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/78248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80642 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39183 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36266 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17670 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.no-llint (failure)") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3149,6 +3176,20 @@ template <class TreeBuilder> TreeClassExpression Parser<LexerType>::parseClass(T
+
+            // property = context.createProperty(ident, type, SuperBinding::Needed, InferName::Allowed, tag);
+            
+            dataLogIf(Options::useDebugLog(), "-------- parseClass ClassStaticBlockMode case end\n");
```
We have to consume all the tokens inside static block in order to let the lexer proceed. And here a static block property is created for later use (see the comments in NodesCodegen.cpp below).

hyjorc1 replied:
Currently, I just parse it as a `FuncExprNode`. We could parse it as `BlockNode`. So, `FuncExprNode` vs `BlockNode`?

hyjorc1 replied:
How can we reuse the ASTNode (`FuncExprNode ` or `BlockNode `) in parseClassFieldInitializerSourceElements?

hyjorc1 commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -649,8 +654,9 @@ RegisterID* PropertyListNode::emitBytecode(BytecodeGenerator& generator, Registe
+            dataLogIf(Options::useDebugLog(), "PropertyListNode::emitBytecode isStaticClassElement 1\n");
             staticFieldLocations->append(p->position());
             continue;
         }
```
Here we need to add the positions of static block to `staticFieldLocations` for initializer evaluation.

hyjorc1 commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -719,8 +725,9 @@ RegisterID* PropertyListNode::emitBytecode(BytecodeGenerator& generator, Registe
+                dataLogIf(Options::useDebugLog(), "PropertyListNode::emitBytecode isStaticClassElement 2\n");
                 staticFieldLocations->append(p->position());
                 continue;
             }
```
Ditto

hyjorc1 commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3289,7 +3333,12 @@ template <class TreeBuilder> TreeSourceElements Parser<LexerType>::parseClassFie
+        if (match(OPENBRACE) && isStaticField) {
+            context.appendStatement(sourceElements, parseBlockStatement(context));
+            continue; 
+        }
```
Here, I just parse the static block again which is redundant?

Constellation commented on Source/JavaScriptCore/parser/Nodes.h:
```diff
@@ -2324,15 +2342,20 @@ namespace JSC {
 
     class FuncExprNode : public BaseFuncExprNode {
     public:
+        enum class Type { Normal, StaticBlock };
```
Do we need to define the different enum class here? Can we use the same enum with `FunctionCallValueNode`?

Constellation commented on Source/JavaScriptCore/parser/Nodes.h:
```diff
@@ -975,15 +988,20 @@ namespace JSC {
 
     class FunctionCallValueNode final : public ExpressionNode, public ThrowableExpressionData {
     public:
+        enum class Type { Normal, StaticBlock };
```
Make it enum class Type : uint8_t

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1845,6 +1850,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseTryStatement(
             catchScope->preventVarDeclarations();
             const Identifier* ident = nullptr;
             if (matchSpecIdentifier()) {
+                semanticFailIfTrue(match(AWAIT) && !findClosetFunctionScope(), "Cannot use 'await' as identifier within class static block scope");
```
Is it possible to use `arguments`?

hyjorc1 replied:
Yes, we can use `arguments` under function scope. So, these should have syntax errors:
```
    class C {
        static {
            [arguments]();
        }
    }

    class C {
        static {
            arguments[0];
        }
    }
```
But, these should work.
```
    class C {
        static {
            function inner() {
                [arguments]();
            }
        }
    }

    class C {
        static {
            function inner() {
                arguments[0];
            }
        }
    }

    class C {
        static {
            (a, b) => {
                this.arguments[0];
            };
        }
    }
```

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -897,6 +899,7 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseVariableDecl
 
         failIfTrue(match(PRIVATENAME), "Cannot use a private name to declare a variable");
         if (matchSpecIdentifier()) {
+            failIfTrue(isArgumentsToken() && !findClosetFunctionScope(), "Class static block cannot use 'arguments' as an identifier");
```
Is it possible to use `await`?

hyjorc1 replied:
Yes, we can use `await` key word only under async function but cannot cross static block boundary. So, this should throw syntax error:
```
async function outer() {
    class C {
        static {
            await 0;
        }
    }
}
```
But this should be fine:
```
async function outer() {
    class C {
        static {
            async function inner() {
                await 0;
            }
        }
    }
}
```

And we can only use `await` as identifier under function scope, which is similar to the usage rule of `arguments` in static block. So, this should throw syntax error:
```
class C {
    static {
        let await = 10;
    }
}
```
But, this works.
```
class C {
    static {
        function inner() {
            let await = 10;
        }
    }
}
```

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1890,7 +1896,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseDebuggerState
 
 template <typename LexerType>
-template <class TreeBuilder> TreeStatement Parser<LexerType>::parseBlockStatement(TreeBuilder& context, bool isCatchBlock)
+template <class TreeBuilder> TreeStatement Parser<LexerType>::parseBlockStatement(TreeBuilder& context, bool isCatchBlock, bool isStaticBlock)
```
Instead of passing two boolean, we should pass one enum which indicates the block type, NormalBlock, CatchBlock, and ClassStaticBlock.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4115,6 +4169,9 @@ template <typename TreeBuilder> TreeExpression Parser<LexerType>::parseAssignmen
     TreeExpression lhs = parseConditionalExpression(context);
 
+    if (!lhs && currentScope()->isStaticBlockScope() && match(AWAIT))
+        propagateError();
```
Can you describe what this pattern means via comment?

Constellation replied:
Let's put `UNLIKELY` around this if.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5457,8 +5517,10 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseUnaryExpress
 
-    if (UNLIKELY(match(AWAIT) && (currentFunctionScope()->isAsyncFunctionBoundary() || isModuleParseMode(sourceParseMode()))))
+    if (UNLIKELY(match(AWAIT) && (currentFunctionScope()->isAsyncFunctionBoundary() || isModuleParseMode(sourceParseMode())))) {
+        semanticFailIfTrue(!findClosetAsyncFunctionScope(), "Cannot use 'await' within static block scope");
```
Looks like this is not checking static-block, right?

hyjorc1 replied:
Sorry, the function name `findClosetAsyncFunctionScope ` should be replaced with `findClosetAsyncFunctionScopeUntilStaticBlock`. Please see the examples with class static block, async function, and await above.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3281,6 +3313,9 @@ template <class TreeBuilder> TreeSourceElements Parser<LexerType>::parseClassFie
                 type = DefineFieldNode::Type::ComputedName;
                 break;
             }
+            case OPENBRACE: {
```
What happens if `isStaticField` is false?

hyjorc1 replied:
All syntax errors should be captured in the first parse which is in `parseClass`. Here we just need to assert `isStaticField ` must be true.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3289,17 +3324,36 @@ template <class TreeBuilder> TreeSourceElements Parser<LexerType>::parseClassFie
-        context.appendStatement(sourceElements, defineField);
+            JSTextPosition endPosition = lastTokenEndPosition();
+            TreeArguments arguments = context.createArguments();
+            expression = context.makeStaticBlockFunctionCallNode(startLocation, expression, arguments, endPosition, startPosition, lastTokenEndPosition());
```
Why not construct a node which invokes `this.func()` call here? Then we do not need to have a flag whether this is a static block or not.

hyjorc1 replied:
There are two parsing paths for class definition in current implementation of `Parser.cpp`. The first path is used to check syntax error for the class. And the second path is for parse and emit class field initializers and static blocks.

Constellation commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -1256,11 +1256,18 @@ RegisterID* FunctionCallValueNode::emitBytecode(BytecodeGenerator& generator, Re
+        CallArguments callArguments(generator, nullptr);
+        generator.move(callArguments.thisRegister(), generator.thisRegister());
+        ret = generator.emitCallInTailPosition(returnValue.get(), func.get(), NoExpectedFunction, callArguments, divot(), divotStart(), divotEnd(), DebuggableCall::Yes);
+    } else {
```
I think this is not necessary. See comment in the Parser.cpp.

Constellation commented on Source/JavaScriptCore/parser/Nodes.h:
```diff
@@ -975,15 +988,20 @@ namespace JSC {
 
         ExpressionNode* m_expr;
         ArgumentsNode* m_args;
+        Type m_type;
```
I think this is not necessary, see comment in the Parser.cpp.

Constellation commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -4982,6 +4989,12 @@ void FuncDeclNode::emitBytecode(BytecodeGenerator& generator, RegisterID*)
+        RefPtr<RegisterID> res = generator.emitNewFunctionExpression(generator.finalDestination(dst), this);
+        emitPutHomeObject(generator, res.get(), homeObject.get());
+        return res.get();
+    }
```
I think, instead of having this flag here, we should do this in the caller side of this code generation.

justinmichaud commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1952,6 +1960,38 @@ class Parser {
+        });
+    }
+
+    ALWAYS_INLINE bool findClosetAsyncFunctionScopeUntilStaticBlock()
```
nit: closest

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1902,8 +1907,10 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseBlockStatemen
+        if (type == BlockType::CatchBlock)
             newScope->setIsCatchBlockScope();
+        if (type == BlockType::StaticBlock)
+            newScope->setIsStaticBlockScope();
```
Use `switch (type)` and list up all types.

darinadler commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -1263,6 +1263,23 @@ RegisterID* FunctionCallValueNode::emitBytecode(BytecodeGenerator& generator, Re
+
+RegisterID* StaticBlockFunctionCallNode::emitBytecode(BytecodeGenerator& generator, RegisterID* dst)
+{
+    RefPtr<RegisterID> homeObject = emitHomeObjectForCallee(generator);
```
I suggest we write the type as just `RefPtr` in new code, not `RefPtr<RegisterID>`.

darinadler commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -1263,6 +1263,23 @@ RegisterID* FunctionCallValueNode::emitBytecode(BytecodeGenerator& generator, Re
+RegisterID* StaticBlockFunctionCallNode::emitBytecode(BytecodeGenerator& generator, RegisterID* dst)
+{
+    RefPtr<RegisterID> homeObject = emitHomeObjectForCallee(generator);
+    RefPtr<RegisterID> func = generator.emitNode(m_expr);
```
WebKit coding style generally encourages using words, `function`, rather than word fragments, `func`.

darinadler commented on Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
```diff
@@ -1263,6 +1263,23 @@ RegisterID* FunctionCallValueNode::emitBytecode(BytecodeGenerator& generator, Re
+
+    CallArguments callArguments(generator, nullptr);
+    generator.move(callArguments.thisRegister(), generator.thisRegister());
+    RefPtr<RegisterID> ret = generator.emitCallInTailPosition(returnValue.get(), func.get(), NoExpectedFunction, callArguments, divot(), divotStart(), divotEnd(), DebuggableCall::Yes);
```
WebKit coding style generally encourages using words, `result` or `resultRegister` or `returnValue`, rather than word fragments, `ret`.

darinadler commented on Source/JavaScriptCore/parser/NodeConstructors.h:
```diff
@@ -250,6 +250,17 @@ namespace JSC {
+    inline PropertyNode::PropertyNode(const Identifier& name, Type type, SuperBinding superBinding, ClassElementTag tag)
+        : m_name(&name)
+        , m_expression(nullptr)
+        , m_assign(nullptr)
```
I suggest initializing these in the class definition.

darinadler commented on Source/JavaScriptCore/parser/NodeConstructors.h:
```diff
@@ -250,6 +250,17 @@ namespace JSC {
+        , m_type(type)
+        , m_needsSuperBinding(superBinding == SuperBinding::Needed)
+        , m_classElementTag(static_cast<unsigned>(tag))
+        , m_isOverriddenByDuplicate(false)
```
I suggest initializing this in the class definition.

darinadler commented on Source/JavaScriptCore/parser/Nodes.h:
```diff
@@ -986,6 +999,18 @@ namespace JSC {
+
+        bool isFunctionCall() const final { return true; }
+
+        ExpressionNode* m_expr;
```
I suggest we consider initializing this to nullptr here, even if in practice that never matters.

darinadler commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3281,6 +3312,10 @@ template <class TreeBuilder> TreeSourceElements Parser<LexerType>::parseClassFie
+            case OPENBRACE: {
+                RELEASE_ASSERT(isStaticField);
+                break;
+            }
```
There’s no need for braces here.

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1948,6 +1956,38 @@ class Parser {
         return nullptr;
     }
 
+    ALWAYS_INLINE bool findClosestFunctionScopeUntilStaticBlock()
```
This is called in the hot code and it is not great since this is too costly. Can we have this information in each scope?
For example, if the upper scope has this flag, then the underlying scope should inherit it, unless it is function boundary. We should have a bool flag in scope which says this information, which is configured at the construction of the scope.

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1948,6 +1956,38 @@ class Parser {
+        });
+    }
+
+    ALWAYS_INLINE bool findClosestAsyncFunctionScopeUntilStaticBlock()
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -772,6 +772,8 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseStatementList
         FALLTHROUGH;
     case AWAIT:
     case YIELD: {
+        semanticFailIfTrue(match(YIELD) && m_parserState.isParsingClassStaticBlock, "Cannot use 'yield' within class static block");
```
Is it correct? What happens if we define a function inside static class block and we use yield inside it?

```
class A {
    static {
         function* gen() {
            yield 42;
         }
    }
}
```

hyjorc1 replied:
Yes, it should be correct. Thanks.

Constellation replied:
And let's add tests for this.

hyjorc1 replied:
Yes, it's added (https://github.com/WebKit/WebKit/compare/6d73623adac8b595710847c996069bac59989f31..5b38af5dcf6c4dfe0e38e6c18671872859e737ae#diff-8fb2e50402b3c18fb15d79f69e63a722895422e3d9695aa92622b6d8f8ed3588R159-R167).

Constellation replied:
Let's remove `isParsingClassStaticBlock`. Having this state in a global ParserState is not a great way to handle the issue.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -772,6 +772,8 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseStatementList
     case AWAIT:
     case YIELD: {
+        semanticFailIfTrue(match(YIELD) && m_parserState.isParsingClassStaticBlock, "Cannot use 'yield' within class static block");
+        semanticFailIfTrue(match(AWAIT) && !findClosestAsyncFunctionScopeUntilStaticBlock(), "Cannot use 'await' within static block scope");
```
This `findClosestAsyncFunctionScopeUntilStaticBlock` is too costly, see comments.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4934,13 +5002,15 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
 
         goto identifierExpression;
     case IDENT: {
+        failIfTrue(isArgumentsToken() && !findClosestFunctionScopeUntilStaticBlock(), "Class static block cannot use 'arguments' as an identifier");
```
This `findClosestFunctionScopeUntilStaticBlock` too costly since this is one of the hottest path in parser. See comment around `findClosestFunctionScopeUntilStaticBlock`.

Constellation replied:
Because we already matched `IDENT`, we should do `*m_token.m_data.ident == m_vm.propertyNames->arguments`

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4957,7 +5027,7 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
         }
         if (UNLIKELY(m_parserState.isParsingClassFieldInitializer))
-            failIfTrue(*m_token.m_data.ident == m_vm.propertyNames->arguments, "Cannot reference 'arguments' in class field initializer");
+            failIfTrue(isArgumentsToken(), "Cannot reference 'arguments' in class field initializer");
```
I think we should keep it since we do not need to check `match(IDENT)` again.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -891,6 +893,7 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseVariableDecl
 
         failIfTrue(match(PRIVATENAME), "Cannot use a private name to declare a variable");
         if (matchSpecIdentifier()) {
+            failIfTrue(isArgumentsToken() && !findClosestFunctionScopeUntilStaticBlock(), "Class static block cannot use 'arguments' as an identifier");
```
Let's just check ident with `arguments` directly. We should not call `match(IDENT)` again.

hyjorc1 replied:
There is a chance that it's not `IDENT` type and `matchSpecIdentifier` returns true.
```
    ALWAYS_INLINE bool matchSpecIdentifier()
    {
        return match(IDENT) || isAllowedIdentifierLet(m_token) || isAllowedIdentifierYield(m_token) || isPossiblyEscapedAwait(m_token);
    }
```

Constellation replied:
I think if it is spec identifier, then the token ident exists, so we don't need to check match(IDENT) here anyway

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -2918,11 +2940,13 @@ template <class TreeBuilder> TreeClassExpression Parser<LexerType>::parseClass(T
     // This classHeadScope is similar to functionScope in FunctionExpression with name.
-    AutoPopScopeRef classHeadScope(this, pushScope());
+    bool setIsStaticBlockBoundary = true;
+    AutoPopScopeRef classHeadScope(this, pushScope(setIsStaticBlockBoundary));
```
Why not calling `setIsStaticBlock` in the following sequence? No other `pushScope` is taking parameter for the created scope.

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -805,6 +816,7 @@ struct Scope {
         m_isArrowFunction = false;
         m_isAsyncFunction = false;
         m_isAsyncFunctionBoundary = false;
+        m_isStaticBlockBoundary = false;
```
Why isn't `m_isStaticBlock` set `false` too? A new function's scope should not inherit static-block bits, right?

```
static {
    function test() {
        // this scope is not static block's scope
    }
}
```

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1273,23 +1287,28 @@ class Parser {
             isAsyncFunction = m_scopeStack.last().isAsyncFunction();
+            isStaticBlock = m_scopeStack.last().isStaticBlock();
+            if (setIsStaticBlockBoundary)
+                isStaticBlockBoundary = m_scopeStack.last().isStaticBlockBoundary();
```
I think this should be set by `setIsStaticBlock`, not by this function

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4934,13 +5013,15 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
 
         goto identifierExpression;
     case IDENT: {
+        semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && isArgumentsIdentifier(), "Cannot use 'arguments' as an identifier in static block");
```
Isn't it `isStaticBlock()`? Otherwise,

```
static {
    {
        arguments;
    }
}
```
it does not become error (please add a test for this too).

hyjorc1 replied:
It should have a syntax error. `arguments` only valid under functions if it is wrapped by static block first.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4934,13 +5013,15 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
         return context.createThisExpr(location);
     }
     case AWAIT:
+        semanticFailIfTrue(currentScope()->isStaticBlockBoundary(), "The 'await' keyword is disallowed in the IdentifierReference position within static block");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -772,6 +772,10 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseStatementList
         FALLTHROUGH;
     case AWAIT:
     case YIELD: {
+        if (UNLIKELY(currentScope()->isStaticBlockBoundary())) {
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -891,6 +895,7 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseVariableDecl
 
         failIfTrue(match(PRIVATENAME), "Cannot use a private name to declare a variable");
         if (matchSpecIdentifier()) {
+            semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && isArgumentsIdentifier(), "Cannot use 'arguments' as an identifier in static block");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1618,6 +1623,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseBreakStatemen
     JSTextPosition end = tokenEndPosition();
     next();
     
+    semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && !currentScope()->breakIsValid(), "'break' cannot cross static block boundary");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1640,6 +1646,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseContinueState
     JSTextPosition end = tokenEndPosition();
     next();
     
+    semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && !currentScope()->continueIsValid(), "'continue' cannot cross static block boundary");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1660,7 +1667,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseReturnStateme
     ASSERT(match(RETURN));
     JSTokenLocation location(tokenLocation());
-    semanticFailIfFalse(currentScope()->isFunction(), "Return statements are only valid inside functions");
+    semanticFailIfFalse(currentScope()->isFunction() && !currentScope()->isStaticBlockBoundary(), "Return statements are only valid inside functions");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -1833,6 +1840,7 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseTryStatement(
         } else {
             handleProductionOrFail(OPENPAREN, "(", "start", "'catch' target");
             DepthManager statementDepth(&m_statementDepth);
+            semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && match(AWAIT), "Cannot use 'await' as identifier within static block");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -2950,7 +2974,7 @@ template <class TreeBuilder> TreeClassExpression Parser<LexerType>::parseClass(T
     consumeOrFail(OPENBRACE, "Expected opening '{' at the start of a class body");
 
-    AutoPopScopeRef classScope(this, pushScope());
+    AutoPopScopeRef classScope(this, pushScope(setIsStaticBlockBoundary));
```
Why not calling `setStaticBlock` after `pushScope()` instead of passing this parameter? The other scope creation is not doing this.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3047,6 +3074,7 @@ template <class TreeBuilder> TreeClassExpression Parser<LexerType>::parseClass(T
             break;
         case OPENBRACKET:
             next();
+            semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && match(IDENT) && isArgumentsIdentifier(), "Cannot use 'arguments' as an identifier in static block");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.h:
```diff
@@ -1874,7 +1894,7 @@ class Parser {
     ALWAYS_INLINE bool canUseIdentifierAwait()
     {
-        return m_parserState.allowAwait && !currentScope()->isAsyncFunction() && m_scriptMode != JSParserScriptMode::Module;
+        return m_parserState.allowAwait && !currentScope()->isAsyncFunction() && !currentScope()->isStaticBlockBoundary() && m_scriptMode != JSParserScriptMode::Module;
```
Ditto. (it should be `isStaticBlock`).

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5451,8 +5533,10 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseUnaryExpress
 
-    if (UNLIKELY(match(AWAIT) && (currentFunctionScope()->isAsyncFunctionBoundary() || isModuleParseMode(sourceParseMode()))))
+    if (UNLIKELY(match(AWAIT) && (currentFunctionScope()->isAsyncFunctionBoundary() || isModuleParseMode(sourceParseMode())))) {
+        semanticFailIfTrue(currentScope()->isStaticBlockBoundary(), "Cannot use 'await' within static block");
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -4109,6 +4172,22 @@ template <typename TreeBuilder> TreeExpression Parser<LexerType>::parseAssignmen
+    // parseAssignmentExpression. Therefore, we need to capture the error directly after
+    // parseConditionalExpression. Besides, the usage of `await` is strictly limited in 
+    // class static block.
+    if (UNLIKELY(!lhs && currentScope()->isStaticBlockBoundary() && match(AWAIT)))
```
Ditto.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -2918,11 +2940,13 @@ template <class TreeBuilder> TreeClassExpression Parser<LexerType>::parseClass(T
     classHeadScope->preventVarDeclarations();
     classHeadScope->setStrictMode();
     next();
+    semanticFailIfTrue(currentScope()->isStaticBlockBoundary() && match(AWAIT), "Cannot use 'await' as a class name within static block");
```
Ditto.

Constellation reviewed (changes requested):
Need to discuss offline. I think the current patch is going to the wrong direction about scope information.

Constellation reviewed (commented):
Discussed with @hyjorc1 about how to propagate scope information, and we agree that we need slightly different approach. @hyjorc1 is working on this.

Constellation commented on JSTests/stress/class-static-block.js:
```diff
@@ -0,0 +1,597 @@
+            }
+        }
+    `);
+}, `SyntaxError: 'break' cannot cross static block boundary.`);
```
Can you also test things like this too?
```
while (false) {
    class C {
         static {
            {
                break;
            }
        }
    }
}
```

Constellation commented on JSTests/stress/class-static-block.js:
```diff
@@ -0,0 +1,597 @@
+            }
+        }
+    `);
+}, `SyntaxError: 'break' cannot cross static block boundary.`);
```
Ditto Can you also test things like this too?
```
    class C {
         static {
            {
                break;
            }
        }
    }
```

Constellation commented on JSTests/stress/class-static-block.js:
```diff
@@ -0,0 +1,597 @@
+            }
+        }
+    `);
+}, `SyntaxError: 'continue' cannot cross static block boundary.`);
```
Ditto for continue too.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Implement support for class static initialization blocks
https://bugs.webkit.org/show_bug.cgi?id=235085
rdar://99056882

Reviewed by Yusuke Suzuki.

Class static initialization block is a new feature of a class to perform
additional static initialization during class definition evaluation.

```
class C {
    static { /* … */ }
}
```

TC39 Spec: https://tc39.es/proposal-class-static-block/
TC39 Proposal: https://github.com/tc39/proposal-class-static-block
MDN Web Doc: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_static_initialization_blocks

In this patch, static blocks are implemented as functions which are
evaluated along with the initialization of static class fields during
class definition evaluation. This can be further optimized by inlining
static block functions to the field initialization.

* JSTests/stress/class-static-block.js: Added.
(assert):
(A):
(assert.C):
(assert.B):
(assert.D):
(assert.A):
(assert.A.friendA.prototype.getX):
(assert.A.friendA.prototype.setX):
(assert.A.prototype.getX):
(assert.inner):
(catch.C.prototype.async inner):
(catch.C):
(catch):
(async inner.C.prototype.async inner):
(async inner.C):
(async inner):
(C.inner):
(C):
(await.C.inner):
(await.C):
(await):
(arguments.C.inner):
(arguments.C):
(arguments):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::PropertyListNode::emitBytecode):
(JSC::FunctionCallValueNode::emitBytecode):
(JSC::FuncExprNode::emitBytecode):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createFunctionExpr):
(JSC::ASTBuilder::createProperty):
(JSC::ASTBuilder::makeFunctionCallNode):
* Source/JavaScriptCore/parser/NodeConstructors.h:
(JSC::PropertyNode::PropertyNode):
(JSC::FunctionCallValueNode::FunctionCallValueNode):
(JSC::FuncExprNode::FuncExprNode):
* Source/JavaScriptCore/parser/Nodes.h:
(JSC::FuncExprNode::isStaticBlockFunction const):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::isArrowFunctionParameters):
(JSC::Parser<LexerType>::parseStatementListItem):
(JSC::Parser<LexerType>::parseVariableDeclarationList):
(JSC::Parser<LexerType>::parseBreakStatement):
(JSC::Parser<LexerType>::parseContinueStatement):
(JSC::Parser<LexerType>::parseReturnStatement):
(JSC::Parser<LexerType>::parseTryStatement):
(JSC::Parser<LexerType>::parseBlockStatement):
(JSC::stringArticleForFunctionMode):
(JSC::stringForFunctionMode):
(JSC::Parser<LexerType>::parseFunctionParameters):
(JSC::Parser<LexerType>::parseFunctionInfo):
(JSC::Parser<LexerType>::parseClass):
(JSC::Parser<LexerType>::parseClassFieldInitializerSourceElements):
(JSC::Parser<LexerType>::parseAssignmentExpression):
(JSC::Parser<LexerType>::parsePrimaryExpression):
(JSC::Parser<LexerType>::parseMemberExpression):
(JSC::Parser<LexerType>::parseUnaryExpression):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::setSourceParseMode):
(JSC::Scope::setIsStaticBlockScope):
(JSC::Scope::isStaticBlockScope):
(JSC::Parser::canUseIdentifierAwait):
(JSC::Parser::disallowedIdentifierAwaitReason):
(JSC::Parser::findClosetFunctionScope):
(JSC::Parser::findClosetAsyncFunctionScope):
(JSC::Parser::findScopeUntilStaticBlock):
* Source/JavaScriptCore/parser/ParserModes.h:
(JSC::isFunctionParseMode):
(JSC::isMethodParseMode):
* Source/JavaScriptCore/parser/SyntaxChecker.h:
(JSC::SyntaxChecker::makeFunctionCallNode):
(JSC::SyntaxChecker::createFunctionExpr):
(JSC::SyntaxChecker::createProperty):
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::toStringSlow):

Canonical link: https://commits.webkit.org/255173@main
